File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        """
        find pattern in either of the string
        """
        pattern = None
        for i in range(0,int(len(str1)/2)+1):
            current = str1[0:i+1]
            patternMod = len(str1)%len(current)
            patternDic = int(len(str1)/len(current))
            if patternMod==0 and current*patternDic==str1:
                pattern = str1[0:i+1]
                break
        if pattern is None:
            return ""
        else:
            largestPattern = None
            initPattern = pattern
            start = 1
            while len(pattern)<=len(str1):
                if pattern*int(len(str1)/len(pattern))==str1 and pattern*int(len(str2)/len(pattern))==str2:
                    largestPattern = pattern
                pattern+=initPattern
            if largestPattern is None:
                return ""
            return largestPattern
        return ""
                
                

File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.cpp
class Solution {
public:
    string gcdOfStrings(string str1, string str2) {
        string divisor = "";
        int currentIndex = 0;
        while( divisor.size()<str1.size()){
            divisor+=str1[currentIndex];
            currentIndex+=1;
            if(str1.size() % divisor.size()==0){
                string tempDivisor = divisor;
                while(tempDivisor.size()<str1.size()){
                    tempDivisor+=divisor;
                }
                if(tempDivisor==str1){
                    break;
                }
            }
        }
        string str2Check = divisor;
        while(str2Check.size()<str2.size()){
            str2Check+=divisor;
        }
        if( str2Check!=str2){
            return "";
        }
        string maximumStr = str1.size()>str2.size()?str1:str2;
        string tempDivisor = divisor;
        string curr = divisor;
        while(tempDivisor.size()<=maximumStr.size()){ 
            if(str1.size()%tempDivisor.size()==0 && str2.size()%tempDivisor.size()==0){
                curr = tempDivisor;
            }
           tempDivisor+=divisor; 
        }
        return curr;
    }
};